contracts/ValidatorSet.sol

```diff
@@ -217,69 +264,77 @@ contract ValidatorSet is IValidatorSet, System, IParamSubscriber {
   function misdemeanor(address validator) external override onlySlash {
-    uint256 index = currentValidatorSetMap[validator];
-    if (index == 0) {
+    uint256 indexPlus1 = currentValidatorSetMap[validator];
+    if (indexPlus1 == 0) {
       return;
     }
-    // the actually index
-    index = index - 1;
-    uint256 income = currentValidatorSet[index].income;
-    currentValidatorSet[index].income = 0;
+    uint index_ = indexPlus1 - 1;
+    uint256 income = currentValidatorSet[index_].income;
+    currentValidatorSet[index_].income = 0;
     uint256 rest = currentValidatorSet.length - 1;
-    address operateAddress = currentValidatorSet[index].operateAddress;
+    address operateAddress = currentValidatorSet[index_].operateAddress;
     emit validatorMisdemeanor(operateAddress, income);
     if (rest == 0) {
       // should not happen, but still protect
       return;
     }
     uint256 averageDistribute = income / rest;
     if (averageDistribute != 0) {
-      for (uint256 i = 0; i < index; i++) {
+      for (uint256 i = 0; i < index_; i++) {
         currentValidatorSet[i].income += averageDistribute;
       }
       uint256 n = currentValidatorSet.length;
-      for (uint256 i = index + 1; i < n; i++) {
+      for (uint256 i = index_ + 1; i < n; i++) {
         currentValidatorSet[i].income += averageDistribute;
       }
     }
   }
 
-  /// Slash the validator for felony behaviors
-  /// @param validator The validator to slash
-  /// @param felonyRound The number of rounds to jail
-  /// @param felonyDeposit The amount of deposits to slash
-  function felony(address validator, uint256 felonyRound, uint256 felonyDeposit) external override onlySlash {
-    uint256 index = currentValidatorSetMap[validator];
-    if (index == 0) {
-      return;
-    }
-    // the actually index
-    index = index - 1;
-    uint256 income = currentValidatorSet[index].income;
+/* @product Called by the Slash contract (only) to slash validators for felony behaviors
+   @param validator: The validator to slash
+   @param felonyRound: The number of rounds to jail the validator
+   @param felonyDeposit: The amount of deposits to slash
+
+   @logic
+        1. If the 'bad' validator is the only validator then he will not be jailed, but
+           his income will be zeroed
+        2. Else the validator will be removed from the current.validators list
+        3. And a sum equal to the 'bad' validator income will be equally divided between the
+           rest of the validators without clearing of the 'bad' validator's income @openissue
+        4. Finally the CandidateHub's jailValidator() function will be invoked to place the
+           validator in jail for felonyRound and slash some amount of deposits. Read its
+           documentation for more details
+*/
+  function felony(address validator, uint256 felonyRound, uint256 felonyDeposit)
+            external override onlySlash onlyIfCurrentValidator(validator) {
```

Comment:
the same reason with `return logic` and modifier, we are expected `return logic`

Reply:
does your comment about return-rather-than-revert logic here applies for both onlySlash onlyIfCurrentValidator modifiers?